### PR TITLE
Make custom autoencoders work with SD2 and SDXL models.

### PR DESCRIPTION
**Pull request by coryMosaicML**
This PR contains a number of QOL changes to make it easier to plug in a custom autoencoder into the existing SD2 and SDXL models. 

Changes:
- AutoEncoder config is now saved as part of the state dict in checkpointing
- Added a `load_autoencoder` function for loading an AutoEncoder object from a composer checkpoint
- Added `LogLatentStatistics` callback for tracking mean and std. dev. of latent channels to ease scaling of the latents
- AutoEncoder now has a `device` function as expected by SD2 and SDXL composer models
- AutoEncoder now returns a latent distribution object as expected by SD2 and SDXL composer models
- SD2 and SDXL models now have the option to download and use custom AutoEncoders by setting `autoencoder_path`
- SD2 and SDXL models now have configurable latent scales and downsampling factors, which may differ with different autoencoders

### CODE REVIEW

File: diffusion/callbacks/log_latent_statistics.py
Comment (Skylion007) [inline diffusion/callbacks/log_latent_statistics.py:24]:
```suggestion
        self.latent_statistics = defaultdict(float)
```

File: diffusion/models/autoencoder.py
Comment (Skylion007) [inline diffusion/models/autoencoder.py:380]: Why not just construct this config sttically in one go?

File: diffusion/callbacks/log_latent_statistics.py
Comment (Skylion007) [inline diffusion/callbacks/log_latent_statistics.py:46]:
```suggestion
        self.latent_statistics = defaultdict(float)
```

File: diffusion/models/models.py
Comment (Skylion007) [inline diffusion/models/models.py:354]: Shouldn't this be a hasattr check instead of autoencoder_path is none check? I don't follow.

File: diffusion/models/autoencoder.py
Comment (Landanjs) [inline diffusion/models/autoencoder.py:730]: Nit
```suggestion
        latent_statistics = {'latent_channel_means': [], 'latent_channel_stds': []}
```

File: diffusion/models/models.py
Comment (Landanjs) [inline diffusion/models/models.py:40]: Why not?
```suggestion
    latent_scale: Optional[float] = 0.18215,
```

File: diffusion/models/models.py
Comment (Landanjs) [inline diffusion/models/models.py:115]: Do the weights need to specified as `torch_dtype=torch.float16` somewhere for `encode_latents_in_fp16`?

File: diffusion/models/models.py
Comment (Landanjs) [reply diffusion/models/models.py:40]: Ah, I see.. to have an option to use `latent_statistics` value. But should this be made an explicit argument?  At the expense of another argument

File: diffusion/models/models.py
Comment (Landanjs) [reply diffusion/models/models.py:354]: +1 hasattr check would be better

**Review by Landanjs — COMMENTED** [review]
Great work!! Mostly nits and confusion about certain decisions.

2 bigish things:
1. There might be a bug in the `GaussianDistribution` class
2. Is encoding the latents as fp16 missing when loading from the custom AE?

File: diffusion/callbacks/log_latent_statistics.py
Comment (coryMosaicML) [reply diffusion/callbacks/log_latent_statistics.py]: Fixed!

File: diffusion/callbacks/log_latent_statistics.py
Comment (coryMosaicML) [reply diffusion/callbacks/log_latent_statistics.py]: Fixed!

File: diffusion/callbacks/log_latent_statistics.py
Comment (coryMosaicML) [reply diffusion/callbacks/log_latent_statistics.py]: It turns out that it isn't even necessary to specify this at all, ever. So, it's gone now.

File: diffusion/models/autoencoder.py
Comment (coryMosaicML) [reply diffusion/models/autoencoder.py]: Agreed. Fixed it.

File: diffusion/models/models.py
Comment (coryMosaicML) [reply diffusion/models/models.py]: Agreed, it's a `hasattr` now.

File: diffusion/models/autoencoder.py
Comment (coryMosaicML) [reply diffusion/models/autoencoder.py:272]: It is used in our generate function to decode from generated latents [here](https://github.com/mosaicml/diffusion/blob/90b97ed93e42cde98bc536501fd9ef16bb9a9ced/diffusion/models/stable_diffusion.py#L541C17-L541C17). I agree it is a bit weird, and we could modify our `generate` to not use `sample` like this. But this is to make things a bit more compatible with the HF autoencoder. There _could_ be some randomness here in sampling from the VAE decoder, but it would degrade performance.

File: diffusion/models/autoencoder.py
Comment (coryMosaicML) [reply diffusion/models/autoencoder.py:371]: If I understand correctly, PyTorch [expects it](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.set_extra_state).

File: diffusion/models/autoencoder.py
Comment (coryMosaicML) [reply diffusion/models/autoencoder.py:651]: Somewhat of an arbitrary choice, but the quantity with the square root is closer to the value needed for normalizing latents later on. It still isn't quite right though because this gives the mrs instead of the rms. This logging isn't useful if one runs with the `log_latent_statistics` callback, so I could remove it if you think it is misleading.

File: diffusion/models/autoencoder.py
Comment (coryMosaicML) [reply diffusion/models/autoencoder.py]: Yes, it was a bug. Should be fixed now!

File: diffusion/models/autoencoder.py
Comment (coryMosaicML) [reply diffusion/models/autoencoder.py]: Should return `self`.

File: diffusion/models/autoencoder.py
Comment (coryMosaicML) [reply diffusion/models/autoencoder.py]: Fixed!

File: diffusion/models/models.py
Comment (coryMosaicML) [reply diffusion/models/models.py:40]: Would you be ok with leaving it as `Optional` but having the default be the `0.18215` value? I think that would allow a user to explicitly set `latent_scale = None` to attempt to pick up the `latent_statistics` value.

File: diffusion/models/autoencoder.py
Comment (Landanjs) [reply diffusion/models/autoencoder.py:273]: Thanks!

File: diffusion/models/autoencoder.py
Comment (Landanjs) [reply diffusion/models/autoencoder.py:372]: Ah, I missed that this was part of a PyTorch `Module`

File: diffusion/models/autoencoder.py
Comment (Landanjs) [reply diffusion/models/autoencoder.py:652]: Sorry if I'm missing something, but couldn't this be an `torch.abs` instead of `torch.sqrt(torch.square())`. It doesn't change the code functionally, but the current implementation seems unnecessary to me.

File: diffusion/models/models.py
Comment (Landanjs) [reply diffusion/models/models.py:40]: I think that would be better.

I still don't like the extra `if latent_scale is None: latent_scale = 0.18215`, but I don't think there is a way to avoid it without adding another argument so it's okay.

File: diffusion/models/models.py
Comment (coryMosaicML) [reply diffusion/models/models.py:117]: Added an option to load custom autoencoder weights in fp16 for when this option is enabled.

File: diffusion/models/autoencoder.py
Comment (coryMosaicML) [reply diffusion/models/autoencoder.py]: Ah, I see what you mean now! Yes, it could just be a `torch.abs`. But, given how the mean metric updates stats, I think it is actually better to leave this as a `torch.square` without the `torch.sqrt`.

File: diffusion/models/models.py
Comment (coryMosaicML) [reply diffusion/models/models.py:40]: Hmm, if you like, the `if latent_scale is None: latent_scale = 0.18215` code could be removed in `model.py` since the `StableDiffusion` class also tries to default to the correct value in the event that `latent_scale` is `None` [here](https://github.com/mosaicml/diffusion/blob/c398359ea28ed0ffd5d7a36723de3b4cb16265fc/diffusion/models/stable_diffusion.py#L117). Though this is just the same ugliness in a different place. 

I'd really like to avoid having to introduce an extra argument for this in any case. What about allowing `latent_scale` to take on a value of `'auto'` to try use the value logged in the autoencoder checkpoint, and leaving the default value the correct number for the base model?

Comment (coryMosaicML) [reply diffusion/models/models.py:40]: I was thinking that in the case of `auto`, we'd only allow `latent_scale` to be a `float` (defaulting to the correct value) or a `str`, and throw an error if it's `auto` without an autoencoder checkpoint to load stats from. So, the check for `None` could be removed. I think a setting for `latent_statistics` would work similarly. Definitely open to making it more specific. I'd rather get this right the first time than having something else to clean up later.

Comment (Landanjs) [reply diffusion/models/models.py:40]: To me, `'latent_statistics'` is a lot clearer than `'auto'`

Comment (coryMosaicML) [reply diffusion/models/models.py]: Ok, I made this change!

File: diffusion/models/autoencoder.py
Comment (coryMosaicML) [reply diffusion/models/autoencoder.py:391]: Yes, they do. Do you think it's needed here? I was aiming for the minimal implementation that's compatible with our current model classes rather than compatibility with HF. A cleaner option could be to change our model classes to use the HF `return_dict` and then switch this back to dict outputs.

**Review by A-Jacobson — APPROVED** [review]
Seems that you've addressed aaron and landans requests already. Approving and will cut a new version once this is merged.

**Pull request merged**
